create-container.c

```diff
@@ -1247,6 +1271,215 @@ int main(int argc, char* argv[])
+		memset(c_v3->hw_pkey_a, 0, sizeof(ecc_key_t));
+		memset(c_v3->hw_pkey_d, 0, sizeof(mldsa_key_t));
+		if (params.hw_keyfn_a) {
+			getPublicKeyRaw(&pubkeyraw, params.hw_keyfn_a);
+			verbose_print((char *) "pubkey A = ", pubkeyraw, sizeof(pubkeyraw));
+			memcpy(c_v3->hw_pkey_a, pubkeyraw, sizeof(ecc_key_t));
+		}
+		if (params.hw_keyfn_d) {
+			size_t sLen = sizeof(c_v3->hw_pkey_d);
+			int r = readBinaryFile(c_v3->hw_pkey_d, &sLen,params.hw_keyfn_d);
+			if (0 != r || sLen != DILITHIUM_PUB_KEY_LENGTH)
+				die(EX_SOFTWARE, "Failure reading HW PUBKEY D : %s",params.hw_keyfn_d);
+			verbose_print((char *) "pubkey D = ", c_v3->hw_pkey_d, sizeof(c_v3->hw_pkey_d));
+		}
+		p = sha3_512(c_v3->hw_pkey_a, sizeof(ecc_key_t) + sizeof(mldsa_key_t), md);
+		if (!p)
+			die(EX_SOFTWARE, "%s", "Cannot get SHA3-512");
+		verbose_print((char *) "HW keys hash = ", md, sizeof(md));
+
+		ph_v3 = (ROM_prefix_header_v3_raw*)&(c_v3->prefix);
+		ph_v3->ver_alg.version = cpu_to_be16(3);
+		ph_v3->ver_alg.hash_alg = 3;
+		ph_v3->ver_alg.sig_alg = 3;
+		ph_v3->reserved = 0;
+
+		// Set flags.
+		if (params.hw_flags) {
+			if (!isValidHex(params.hw_flags, 4))
+				die(EX_DATAERR, "%s",
+				    "Invalid input for hw-flags, expecting a 4 byte hexadecimal value");
+			uint32_t data;
+			sscanf(params.hw_flags, "%x", &data);
+			ph_v3->flags = cpu_to_be32(data);
+			verbose_msg("hw-flags = %#010x", data);
+		} else {
+			ph_v3->flags = cpu_to_be32(0x80000000);
+		}
+		memset(ph_v3->payload_hash, 0, sizeof(sha2_hash_t));
+		memset(ph_v3->ecid, 0, ECID_SIZE);
+		memset(ph_v3->reserved2, 0, sizeof(ph_v3->reserved2));
+
+		pd_v3 = (ROM_prefix_data_v3_raw*)&c_v3->prefix_data;
+		memset(pd_v3->hw_sig_a, 0, sizeof(ecc_signature_t));
+		memset(pd_v3->hw_sig_d, 0, sizeof(mldsa_signature_t));
+
+		// Write the HW signatures.
+		if (params.hw_sigfn_a) {
+			getSigRaw(&sigraw, params.hw_sigfn_a);
+			verbose_print((char *) "signature A = ", sigraw, sizeof(sigraw));
+			memcpy(pd_v3->hw_sig_a, sigraw, sizeof(ecc_key_t));
+		}
+		if (params.hw_sigfn_d) {
+			size_t sLen = sizeof(pd_v3->hw_sig_d);
+			int r = readBinaryFile(pd_v3->hw_sig_d, &sLen,params.hw_sigfn_d);
+			if (0 != r || sLen != MLDSA_87_SIG_LENGTH)
+				die(EX_SOFTWARE, "Failure reading HW SIG D : %s",params.hw_sigfn_d);
+			verbose_print((char *) "signature D = ", pd_v3->hw_sig_d, sizeof(pd_v3->hw_sig_d));
+		}
+		memset(pd_v3->sw_pkey_p, 0, sizeof(ecc_key_t));
+		memset(pd_v3->sw_pkey_s, 0, sizeof(mldsa_key_t));
+
+		// Write the FW keys.
+		if (params.sw_keyfn_p) {
+			getPublicKeyRaw(&pubkeyraw, params.sw_keyfn_p);
+			verbose_print((char *) "pubkey P = ", pubkeyraw, sizeof(pubkeyraw));
+			memcpy(pd_v3->sw_pkey_p, pubkeyraw, sizeof(ecc_key_t));
+			ph_v3->sw_key_count++;
+			ph_v3->payload_size += sizeof(ecc_key_t);
+		}
+		if (params.sw_keyfn_s) {
+			size_t sLen = sizeof(pd_v3->sw_pkey_s);
+			int r = readBinaryFile(pd_v3->sw_pkey_s, &sLen,params.sw_keyfn_s);
+			if (0 != r || sLen != DILITHIUM_PUB_KEY_LENGTH)
```

Comment:
I like to mark things like this where this needs to change to MLDSA as todos so we remember later

Reply:
Done